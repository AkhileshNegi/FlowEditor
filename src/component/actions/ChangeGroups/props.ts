import { ChangeGroups } from '../../../flowTypes';
import { Type } from '../../../config';

export default interface ChangeGroupsFormProps {
    typeConfig: Type;
    action: ChangeGroups;
    updateAction: (action: ChangeGroups) => void;
    onBindWidget: (ref: any) => void;
    removeWidget: (name: string) => void;
};
