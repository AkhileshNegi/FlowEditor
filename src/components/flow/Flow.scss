@import '~/variables.scss';

.loading {
    visibility: hidden;
    height: inherit;
}

.loaded {
    height: inherit;
    position: relative;
}

.nodeList {
    position: relative;
    user-select: none;
}

.dragSelection {
    position: absolute;
    background: red;
    border: 0px solid darken($selection_square, 10%);
    opacity: 0;
    pointer-events: none;
    z-index: $z_selection;
}
