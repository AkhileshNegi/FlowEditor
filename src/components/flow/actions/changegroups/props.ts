import { ActionFormHelper, Type } from '~/config/typeConfigs';
import { ChangeGroups } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ChangeGroupsFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    formHelper: ActionFormHelper;
    typeConfig: Type;

    // update handlers
    updateAction(action: ChangeGroups): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}
export default ChangeGroupsFormProps;
