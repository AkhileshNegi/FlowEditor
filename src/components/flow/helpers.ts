import { Type, Types } from '~/config/interfaces';
import { getType, getTypeConfig } from '~/config/typeConfigs';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const determineTypeConfig = (nodeSettings: NodeEditorSettings): Type => {
    const renderNode = nodeSettings.originalNode;
    const node = renderNode && renderNode.node;

    if (nodeSettings.originalAction && nodeSettings.originalAction.type) {
        return getTypeConfig(nodeSettings.originalAction.type);
    } else if (node && node.actions && node.actions.length > 0) {
        return getTypeConfig(node.actions[node.actions.length - 1].type);
    } else {
        try {
            const type = getType(renderNode);
            const config = getTypeConfig(type);
            if (config.type !== Types.missing) {
                return config;
            }
            // tslint:disable-next-line:no-empty
        } catch (Error) {}
    }

    // Account for ghost nodes
    if (node && node.router) {
        return getTypeConfig(node.router.type);
    }

    throw new Error(`Couldn't determine type config for: ${node.uuid}`);
};
