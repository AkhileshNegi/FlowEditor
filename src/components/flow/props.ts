import { UpdateLocalizations } from '~/components/nodeeditor/NodeEditor';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { AssetStore, RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';
import { DispatchWithState, GetState } from '~/store/thunks';

export interface ActionFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;
    assets: AssetStore;

    // update handlers
    updateAction(
        action: AnyAction,
        onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
    ): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface RouterFormProps {
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    assets: AssetStore;

    // update handlers
    updateRouter(renderNode: RenderNode): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
    language: Asset;
    nodeSettings: NodeEditorSettings;
    updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
    onClose(canceled: boolean): void;
}
