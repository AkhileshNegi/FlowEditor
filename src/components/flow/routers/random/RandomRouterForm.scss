@import '~/variables.scss';

.lead-in {
    margin-bottom: 5px;
    display: inline-block;
    vertical-align: top;
    margin-top: 6px;
}


.bucket_select {
    width: 154px;
    display: inline-block;
    margin-left: 10px;
    margin-bottom: -10px;
    vertical-align: top;
}

$bucketSpacing: 6px;

.bucket_list {
    display: flex;
    flex-wrap: wrap;
    padding: 10px 0;
    margin: 0px -$bucketSpacing;
    margin-top: 5px;
}

.bucket_name {
    flex: 1 1 120px;
    margin: $bucketSpacing;
}