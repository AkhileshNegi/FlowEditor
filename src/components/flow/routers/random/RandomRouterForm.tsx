import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet } from '~/components/dialog/Dialog';
import { RouterFormProps } from '~/components/flow/props';
import OptionalTextInput from '~/components/form/optionaltext/OptionalTextInput';
import SelectElement, { SelectOption } from '~/components/form/select/SelectElement';
import TextInputElement from '~/components/form/textinput/TextInputElement';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Exit } from '~/flowTypes';
import { FormState, mergeForm, SelectOptionEntry, StringEntry } from '~/store/nodeEditor';
import { small } from '~/utils/reactselect';

import { BUCKET_OPTIONS, fillOutExits, nodeToState, stateToNode } from './helpers';
import * as styles from './RandomRouterForm.scss';

// TODO: Remove use of Function
// tslint:disable:ban-types
export enum InputToFocus {
    args = 'args',
    min = 'min',
    max = 'max',
    exit = 'exit'
}

export interface RandomRouterFormState extends FormState {
    bucketChoice: SelectOptionEntry;
    resultName: StringEntry;
    exits: Exit[];
}

export const leadInSpecId = 'lead-in';

export default class RandomRouterForm extends React.Component<
    RouterFormProps,
    RandomRouterFormState
> {
    constructor(props: RouterFormProps) {
        super(props);

        this.state = nodeToState(this.props.nodeSettings);

        bindCallbacks(this, {
            include: [/^on/, /^handle/]
        });
    }

    public static contextTypes = {
        assetService: fakePropType
    };

    private handleUpdateResultName(value: string): void {
        this.setState({ resultName: { value } });
    }

    private handleBucketsChanged(selected: SelectOption): boolean {
        // create new exits if needed

        const count = parseInt(selected.value, 10);

        let exits = this.state.exits.concat([]);

        // prune off if we have too many
        exits = exits.slice(0, count);

        // add any that we still need
        exits = fillOutExits(exits, count);

        const updates: Partial<RandomRouterFormState> = {
            bucketChoice: { value: selected }
        };

        const updated = mergeForm(this.state, updates);
        this.setState({ ...updated, exits });

        return updated.valid;
    }

    private handleSave(): void {
        this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
        this.props.onClose(false);
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave, disabled: !this.state.valid },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    private handleBucketNameChanged(exit: Exit, value: string): void {
        const exits = this.state.exits;
        exits.find((ex: Exit) => ex.uuid === exit.uuid).name = value;
        this.setState({ exits });
    }

    private renderBucketNames(): any {
        return this.state.exits.map((exit: Exit) => (
            <TextInputElement
                key={exit.uuid}
                __className={styles.bucketName}
                data-spec="optional-field"
                name={exit.uuid}
                entry={{ value: exit.name }}
                onChange={(value: string) => {
                    this.handleBucketNameChanged(exit, value);
                }}
            />
        ));
    }

    public renderEdit(): JSX.Element {
        const typeConfig = this.props.typeConfig;

        const OPTIONS = BUCKET_OPTIONS.concat([]);
        if (BUCKET_OPTIONS.indexOf(this.state.bucketChoice.value) === -1) {
            OPTIONS.push(this.state.bucketChoice.value);
        }

        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />
                <div className={styles.leadIn}>Split them randomly into one of</div>
                <div className={styles.bucketSelect}>
                    <SelectElement
                        styles={small}
                        name="Buckets"
                        entry={this.state.bucketChoice}
                        onChange={this.handleBucketsChanged}
                        options={OPTIONS}
                    />
                </div>
                <div className={styles.bucketList}>{this.renderBucketNames()}</div>
                <OptionalTextInput
                    name="Result Name"
                    value={this.state.resultName}
                    onChange={this.handleUpdateResultName}
                    toggleText="Save as.."
                    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
                />
            </Dialog>
        );
    }

    public render(): JSX.Element {
        return this.renderEdit();
    }
}
