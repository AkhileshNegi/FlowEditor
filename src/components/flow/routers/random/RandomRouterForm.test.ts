import { RouterFormProps } from "components/flow/props";
import RandomRouterForm from "components/flow/routers/random/RandomRouterForm";
import { composeComponentTestUtils, mock } from "testUtils";
import {
  createRandomNode,
  createRenderNode,
  createSendMsgAction,
  getRouterFormProps
} from "testUtils/assetCreators";
import * as utils from "utils";
import { Types } from "config/interfaces";

mock(utils, "createUUID", utils.seededUUIDs());

const { setup } = composeComponentTestUtils<RouterFormProps>(
  RandomRouterForm,
  getRouterFormProps(createRandomNode(2))
);

describe(RandomRouterForm.name, () => {
  it("should render", () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("should initialize existing random", () => {
    const { wrapper } = setup(true, {
      nodeSettings: { $set: { originalNode: createRandomNode(5) } }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should convert from a non-random node", () => {
    const { wrapper, instance, props } = setup(true, {
      nodeSettings: {
        $set: {
          originalAction: createSendMsgAction({ text: "" }),
          originalNode: createRenderNode({
            actions: [createSendMsgAction({ text: "A message" })],
            exits: [{ uuid: utils.createUUID() }],
            ui: {
              type: Types.execute_actions,
              position: { left: 100, top: 100 }
            }
          })
        }
      }
    });

    instance.handleSave();

    // our orginal node should still only have one exit
    expect(props.nodeSettings.originalNode.node.exits.length).toBe(1);

    expect(wrapper).toMatchSnapshot();
    expect(props.updateRouter).toMatchCallSnapshot();
    expect(props.nodeSettings.originalNode).toMatchSnapshot();
  });
});
