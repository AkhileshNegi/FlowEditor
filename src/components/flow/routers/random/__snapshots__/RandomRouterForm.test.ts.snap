// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomRouterForm should convert from a non-random node 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    Split them randomly into one of
  </div>
  <div
    className="bucketSelect"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "2 buckets",
            "value": "2",
          },
        }
      }
      name="Buckets"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "2 buckets",
            "value": "2",
          },
          Object {
            "label": "3 buckets",
            "value": "3",
          },
          Object {
            "label": "4 buckets",
            "value": "4",
          },
          Object {
            "label": "5 buckets",
            "value": "5",
          },
          Object {
            "label": "6 buckets",
            "value": "6",
          },
          Object {
            "label": "7 buckets",
            "value": "7",
          },
          Object {
            "label": "8 buckets",
            "value": "8",
          },
          Object {
            "label": "9 buckets",
            "value": "9",
          },
          Object {
            "label": "10 buckets",
            "value": "10",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <div
    className="bucketList"
  >
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="141e0356-9fdd-47d3-a12d-16efc5b45276"
      name="141e0356-9fdd-47d3-a12d-16efc5b45276"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="e81b3f1a-0863-4418-8a72-9c9f8aa72b7c"
      name="e81b3f1a-0863-4418-8a72-9c9f8aa72b7c"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`RandomRouterForm should convert from a non-random node 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
        },
      ],
      "router": Object {
        "categories": Array [
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "Bucket 1",
            "uuid": "141e0356-9fdd-47d3-a12d-16efc5b45276",
          },
          Object {
            "exit_uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
            "name": "Bucket 2",
            "uuid": "e81b3f1a-0863-4418-8a72-9c9f8aa72b7c",
          },
        ],
        "type": "random",
      },
      "uuid": "f05fa107-03db-4315-a211-53da63603e6f",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_random",
    },
  },
]
`;

exports[`RandomRouterForm should initialize existing random 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    Split them randomly into one of
  </div>
  <div
    className="bucketSelect"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "5 buckets",
            "value": "5",
          },
        }
      }
      name="Buckets"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "2 buckets",
            "value": "2",
          },
          Object {
            "label": "3 buckets",
            "value": "3",
          },
          Object {
            "label": "4 buckets",
            "value": "4",
          },
          Object {
            "label": "5 buckets",
            "value": "5",
          },
          Object {
            "label": "6 buckets",
            "value": "6",
          },
          Object {
            "label": "7 buckets",
            "value": "7",
          },
          Object {
            "label": "8 buckets",
            "value": "8",
          },
          Object {
            "label": "9 buckets",
            "value": "9",
          },
          Object {
            "label": "10 buckets",
            "value": "10",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <div
    className="bucketList"
  >
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="e67ce051-f3e4-4c42-807b-e393df5fa382"
      name="e67ce051-f3e4-4c42-807b-e393df5fa382"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="e0bc3776-792f-49d2-b4be-48be2cf8159a"
      name="e0bc3776-792f-49d2-b4be-48be2cf8159a"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 3",
        }
      }
      key="6c22884b-0e35-4206-982e-18320691eda9"
      name="6c22884b-0e35-4206-982e-18320691eda9"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 4",
        }
      }
      key="a62e8284-49f8-4292-adbc-d5f6607f85c5"
      name="a62e8284-49f8-4292-adbc-d5f6607f85c5"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 5",
        }
      }
      key="0c8c9239-9b59-473a-a229-d8a9693be270"
      name="0c8c9239-9b59-473a-a229-d8a9693be270"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`RandomRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    Split them randomly into one of
  </div>
  <div
    className="bucketSelect"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "2 buckets",
            "value": "2",
          },
        }
      }
      name="Buckets"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "2 buckets",
            "value": "2",
          },
          Object {
            "label": "3 buckets",
            "value": "3",
          },
          Object {
            "label": "4 buckets",
            "value": "4",
          },
          Object {
            "label": "5 buckets",
            "value": "5",
          },
          Object {
            "label": "6 buckets",
            "value": "6",
          },
          Object {
            "label": "7 buckets",
            "value": "7",
          },
          Object {
            "label": "8 buckets",
            "value": "8",
          },
          Object {
            "label": "9 buckets",
            "value": "9",
          },
          Object {
            "label": "10 buckets",
            "value": "10",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <div
    className="bucketList"
  >
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      name="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucketName"
      data-spec="optional-field"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      name="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;
